File: src/services/code_action/autofix_ts_syntax.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

class mapper target_loc kind =
  object (this)
    inherit [Loc.t] Flow_ast_contains_mapper.mapper as super

    method private is_target loc = Loc.equal target_loc loc

    method loc_annot_contains_target loc = Loc.contains loc target_loc

    method! type_ t =
      let open Flow_ast.Type in
      match t with
      | (loc, Unknown comments) when kind = `UnknownType && this#is_target loc ->
        Ast_builder.Types.mixed ?comments ()
      | (loc, Never comments) when kind = `NeverType && this#is_target loc ->
        Ast_builder.Types.empty ?comments ()
      | (loc, Undefined comments) when kind = `UndefinedType && this#is_target loc ->
        Ast_builder.Types.void ?comments ()
      | (loc, Keyof { Keyof.argument; comments }) when kind = `KeyofType && this#is_target loc ->
        let targs = Ast_builder.Types.type_args [super#type_ argument] in
        Ast_builder.Types.unqualified_generic ?comments ~targs "$Keys"
      | (loc, ReadOnly { ReadOnly.argument = (_, Array { Array.argument; _ }); comments })
        when kind = `ReadOnlyArrayType && this#is_target loc ->
        let targs = Ast_builder.Types.type_args [super#type_ argument] in
        Ast_builder.Types.unqualified_generic ?comments ~targs "$ReadOnlyArray"
      | (loc, ReadOnly { ReadOnly.argument = (_, Tuple _) as argument; comments })
        when kind = `ReadOnlyTupleType && this#is_target loc ->
        let targs = Ast_builder.Types.type_args [super#type_ argument] in
        Ast_builder.Types.unqualified_generic ?comments ~targs "$ReadOnly"
      | _ -> super#type_ t

    method! type_param (loc, tparam) =
      let open Flow_ast.Type.TypeParam in
      match tparam with
      | { bound_kind = Extends; _ } when kind = `TypeParamExtends && this#is_target loc ->
        (Loc.none, { tparam with bound_kind = Colon })
      | { variance = Some (v_loc, Flow_ast.Variance.{ kind = InOut; _ }); _ }
        when kind = `InOutVariance && this#is_target v_loc ->
        (Loc.none, { tparam with variance = None })
      | _ -> super#type_param (loc, tparam)

    method! variance variance =
      let open Flow_ast.Variance in
      let (loc, { kind = variance_kind; comments }) = variance in
      match variance_kind with
      | Readonly when kind = `ReadonlyVariance && this#is_target loc ->
        (Loc.none, { kind = Plus; comments })
      | In when kind = `InVariance && this#is_target loc -> (Loc.none, { kind = Minus; comments })
      | Out when kind = `OutVariance && this#is_target loc -> (Loc.none, { kind = Plus; comments })
      | _ -> super#variance variance

    method! expression e =
      let open Flow_ast.Expression in
      match e with
      | (loc, TSTypeCast { TSTypeCast.expression; kind = TSTypeCast.As annot; comments })
        when kind = `AsExpression && this#is_target loc ->
        let expression = super#expression expression in
        let annot = super#type_ annot in
        Ast_builder.Expressions.typecast ?comments expression annot
      | (loc, TSTypeCast { TSTypeCast.expression; kind = TSTypeCast.Satisfies annot; comments })
        when kind = `SatisfiesExpression && this#is_target loc ->
        let expression = super#expression expression in
        let annot = super#type_ annot in
        Ast_builder.Expressions.typecast ?comments expression annot
      | _ -> super#expression e
  end

let convert_unknown_type ast loc =
  let mapper = new mapper loc `UnknownType in
  mapper#program ast

let convert_never_type ast loc =
  let mapper = new mapper loc `NeverType in
  mapper#program ast

let convert_undefined_type ast loc =
  let mapper = new mapper loc `UndefinedType in
  mapper#program ast

let convert_keyof_type ast loc =
  let mapper = new mapper loc `KeyofType in
  mapper#program ast

let convert_type_param_extends ast loc =
  let mapper = new mapper loc `TypeParamExtends in
  mapper#program ast

let convert_readonly_variance ast loc =
  let mapper = new mapper loc `ReadonlyVariance in
  mapper#program ast

let convert_in_variance ast loc =
  let mapper = new mapper loc `InVariance in
  mapper#program ast

let convert_out_variance ast loc =
  let mapper = new mapper loc `OutVariance in
  mapper#program ast

let remove_in_out_variance ast loc =
  let mapper = new mapper loc `InOutVariance in
  mapper#program ast

let convert_as_expression ast loc =
  let mapper = new mapper loc `AsExpression in
  mapper#program ast

let convert_satisfies_expression ast loc =
  let mapper = new mapper loc `SatisfiesExpression in
  mapper#program ast

let convert_readonly_array_type ast loc =
  let mapper = new mapper loc `ReadOnlyArrayType in
  mapper#program ast

let convert_readonly_tuple_type ast loc =
  let mapper = new mapper loc `ReadOnlyTupleType in
  mapper#program ast

File: rust_port/crates/flow_services_code_action/src/autofix_ts_syntax.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use dupe::Dupe;
use flow_parser::ast;
use flow_parser::ast::expression::ExpressionInner;
use flow_parser::ast::types::TypeInner;
use flow_parser::ast_visitor::AstVisitor;
use flow_parser::ast_visitor::TypeParamsContext;
use flow_parser::ast_visitor::map_expression_default;
use flow_parser::ast_visitor::map_program_default;
use flow_parser::ast_visitor::map_statement_default;
use flow_parser::ast_visitor::map_type_default;
use flow_parser::ast_visitor::map_type_param_default;
use flow_parser::ast_visitor::map_variance_default;
use flow_parser::loc::LOC_NONE;
use flow_parser::loc::Loc;
use flow_parser_utils::ast_builder;

use crate::contains_mapper::ContainsMapper;

#[derive(PartialEq, Eq)]
enum Kind {
    UnknownType,
    NeverType,
    UndefinedType,
    KeyofType,
    ReadOnlyArrayType,
    ReadOnlyTupleType,
    TypeParamExtends,
    TypeParamColon,
    InOutVariance,
    ReadonlyVariance,
    InVariance,
    OutVariance,
    AsExpression,
    SatisfiesExpression,
}

struct Mapper {
    contains: ContainsMapper,
    kind: Kind,
}

impl AstVisitor<'_, Loc> for Mapper {
    fn normalize_loc(loc: &Loc) -> &Loc {
        loc
    }

    fn normalize_type(type_: &Loc) -> &Loc {
        type_
    }

    fn map_type_(&mut self, t: &ast::types::Type<Loc, Loc>) -> ast::types::Type<Loc, Loc> {
        match &**t {
            TypeInner::Unknown { loc, comments }
                if self.kind == Kind::UnknownType && self.contains.is_target(loc) =>
            {
                ast_builder::types::mixed(None, comments.clone())
            }
            TypeInner::Never { loc, comments }
                if self.kind == Kind::NeverType && self.contains.is_target(loc) =>
            {
                ast_builder::types::empty(None, comments.clone())
            }
            TypeInner::Undefined { loc, comments }
                if self.kind == Kind::UndefinedType && self.contains.is_target(loc) =>
            {
                ast_builder::types::void(None, comments.clone())
            }
            TypeInner::Keyof { loc, inner }
                if self.kind == Kind::KeyofType && self.contains.is_target(loc) =>
            {
                let targs = ast_builder::types::type_args(
                    None,
                    None,
                    vec![map_type_default(self, &inner.argument)],
                );
                ast_builder::types::unqualified_generic(
                    inner.comments.clone(),
                    None,
                    Some(targs),
                    "$Keys",
                )
            }
            TypeInner::ReadOnly { loc, inner }
                if self.kind == Kind::ReadOnlyArrayType && self.contains.is_target(loc) =>
            {
                if let TypeInner::Array {
                    loc: _,
                    inner: array_inner,
                } = &*inner.argument
                {
                    let targs = ast_builder::types::type_args(
                        None,
                        None,
                        vec![map_type_default(self, &array_inner.argument)],
                    );
                    ast_builder::types::unqualified_generic(
                        inner.comments.clone(),
                        None,
                        Some(targs),
                        "ReadonlyArray",
                    )
                } else {
                    map_type_default(self, t)
                }
            }
            TypeInner::ReadOnly { loc, inner }
                if self.kind == Kind::ReadOnlyTupleType && self.contains.is_target(loc) =>
            {
                if let TypeInner::Tuple { .. } = &*inner.argument {
                    let targs = ast_builder::types::type_args(
                        None,
                        None,
                        vec![map_type_default(self, &inner.argument)],
                    );
                    ast_builder::types::unqualified_generic(
                        inner.comments.clone(),
                        None,
                        Some(targs),
                        "Readonly",
                    )
                } else {
                    map_type_default(self, t)
                }
            }
            _ => map_type_default(self, t),
        }
    }

    fn map_type_param(
        &mut self,
        k: &TypeParamsContext,
        tparam: &ast::types::TypeParam<Loc, Loc>,
    ) -> ast::types::TypeParam<Loc, Loc> {
        use ast::types::type_param::BoundKind;
        match tparam {
            ast::types::TypeParam {
                bound_kind: BoundKind::Extends,
                loc,
                ..
            } if self.kind == Kind::TypeParamExtends && self.contains.is_target(loc) => {
                ast::types::TypeParam {
                    loc: LOC_NONE,
                    bound_kind: BoundKind::Colon,
                    ..tparam.clone()
                }
            }
            ast::types::TypeParam {
                bound_kind: BoundKind::Colon,
                loc,
                ..
            } if self.kind == Kind::TypeParamColon && self.contains.is_target(loc) => {
                ast::types::TypeParam {
                    loc: LOC_NONE,
                    bound_kind: BoundKind::Extends,
                    ..tparam.clone()
                }
            }
            ast::types::TypeParam {
                variance:
                    Some(ast::Variance {
                        loc: v_loc,
                        kind: ast::VarianceKind::InOut,
                        ..
                    }),
                ..
            } if self.kind == Kind::InOutVariance && self.contains.is_target(v_loc) => {
                ast::types::TypeParam {
                    loc: LOC_NONE,
                    variance: None,
                    ..tparam.clone()
                }
            }
            _ => map_type_param_default(self, k, tparam),
        }
    }

    fn map_variance(&mut self, variance: &ast::Variance<Loc>) -> ast::Variance<Loc> {
        let ast::Variance {
            loc,
            kind: variance_kind,
            comments,
        } = variance;
        match variance_kind {
            ast::VarianceKind::Readonly
                if self.kind == Kind::ReadonlyVariance && self.contains.is_target(loc) =>
            {
                ast::Variance {
                    loc: LOC_NONE,
                    kind: ast::VarianceKind::Plus,
                    comments: comments.clone(),
                }
            }
            ast::VarianceKind::In
                if self.kind == Kind::InVariance && self.contains.is_target(loc) =>
            {
                ast::Variance {
                    loc: LOC_NONE,
                    kind: ast::VarianceKind::Minus,
                    comments: comments.clone(),
                }
            }
            ast::VarianceKind::Out
                if self.kind == Kind::OutVariance && self.contains.is_target(loc) =>
            {
                ast::Variance {
                    loc: LOC_NONE,
                    kind: ast::VarianceKind::Plus,
                    comments: comments.clone(),
                }
            }
            _ => map_variance_default(self, variance),
        }
    }

    fn map_expression(
        &mut self,
        expr: &ast::expression::Expression<Loc, Loc>,
    ) -> ast::expression::Expression<Loc, Loc> {
        match &**expr {
            ExpressionInner::AsExpression { loc: _, inner }
                if self.kind == Kind::AsExpression && self.contains.is_target(expr.loc()) =>
            {
                let expression = map_expression_default(self, &inner.expression);
                let annot = map_type_default(self, &inner.annot.annotation);
                ast_builder::expressions::typecast(None, inner.comments.clone(), expression, annot)
            }
            ExpressionInner::TSSatisfies { loc: _, inner }
                if self.kind == Kind::SatisfiesExpression
                    && self.contains.is_target(expr.loc()) =>
            {
                let expression = map_expression_default(self, &inner.expression);
                let mapped_annot = self.map_type_annotation(&inner.annot);
                let annot = mapped_annot.annotation;
                ast_builder::expressions::typecast(None, inner.comments.clone(), expression, annot)
            }
            _ => {
                if self.contains.should_map_expression(expr) {
                    map_expression_default(self, expr)
                } else {
                    expr.dupe()
                }
            }
        }
    }

    fn map_program(&mut self, program: &ast::Program<Loc, Loc>) -> ast::Program<Loc, Loc> {
        if self.contains.should_map_program(program) {
            map_program_default(self, program)
        } else {
            program.clone()
        }
    }

    fn map_statement(
        &mut self,
        stmt: &ast::statement::Statement<Loc, Loc>,
    ) -> ast::statement::Statement<Loc, Loc> {
        if self.contains.should_map_statement(stmt) {
            map_statement_default(self, stmt)
        } else {
            stmt.dupe()
        }
    }
}

pub fn convert_unknown_type(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::UnknownType,
    };
    mapper.map_program(ast)
}

pub fn convert_never_type(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::NeverType,
    };
    mapper.map_program(ast)
}

pub fn convert_undefined_type(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::UndefinedType,
    };
    mapper.map_program(ast)
}

pub fn convert_keyof_type(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::KeyofType,
    };
    mapper.map_program(ast)
}

pub fn convert_type_param_extends(
    ast: &ast::Program<Loc, Loc>,
    loc: Loc,
) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::TypeParamExtends,
    };
    mapper.map_program(ast)
}

pub fn convert_type_param_colon(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::TypeParamColon,
    };
    mapper.map_program(ast)
}

pub fn convert_readonly_variance(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::ReadonlyVariance,
    };
    mapper.map_program(ast)
}

pub fn convert_in_variance(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::InVariance,
    };
    mapper.map_program(ast)
}

pub fn convert_out_variance(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::OutVariance,
    };
    mapper.map_program(ast)
}

pub fn remove_in_out_variance(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::InOutVariance,
    };
    mapper.map_program(ast)
}

pub fn convert_as_expression(ast: &ast::Program<Loc, Loc>, loc: Loc) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::AsExpression,
    };
    mapper.map_program(ast)
}

pub fn convert_satisfies_expression(
    ast: &ast::Program<Loc, Loc>,
    loc: Loc,
) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::SatisfiesExpression,
    };
    mapper.map_program(ast)
}

pub fn convert_readonly_array_type(
    ast: &ast::Program<Loc, Loc>,
    loc: Loc,
) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::ReadOnlyArrayType,
    };
    mapper.map_program(ast)
}

pub fn convert_readonly_tuple_type(
    ast: &ast::Program<Loc, Loc>,
    loc: Loc,
) -> ast::Program<Loc, Loc> {
    let mut mapper = Mapper {
        contains: ContainsMapper::new(loc),
        kind: Kind::ReadOnlyTupleType,
    };
    mapper.map_program(ast)
}
